File: programs/normal/src/math/spot_balance.rs
use crate::error::{ NormalResult, ErrorCode };
use crate::math::casting::Cast;
use crate::math::constants::{
    ONE_YEAR,
    PERCENTAGE_PRECISION,
    SPOT_RATE_PRECISION,
    SPOT_UTILIZATION_PRECISION,
};
use crate::math::safe_math::{ SafeDivFloor, SafeMath };
use crate::state::oracle::{ OraclePriceData, StrictOraclePrice };
use crate::state::market::{ BalanceType, Market };
use crate::state::user::SpotPosition;

pub fn get_spot_balance(token_amount: u128, market: &Market, round_up: bool) -> NormalResult<u128> {
    let precision_increase = (10_u128).pow((19_u32).safe_sub(market.decimals)?);

    let mut balance = token_amount.safe_mul(precision_increase)?;

    if round_up && balance != 0 {
        balance = balance.safe_add(1)?;
    }

    Ok(balance)
}

pub fn get_token_amount(balance: u128, market: &Market) -> NormalResult<u128> {
    let precision_decrease = (10_u128).pow((19_u32).safe_sub(market.decimals)?);

    let token_amount = balance.safe_div(precision_decrease)?;

    Ok(token_amount)
}

pub fn get_signed_token_amount(token_amount: u128) -> NormalResult<i128> {
    token_amount.cast()
}

pub fn get_interest_token_amount(
    balance: u128,
    market: &Market,
    interest: u128
) -> NormalResult<u128> {
    let precision_decrease = (10_u128).pow((19_u32).safe_sub(market.decimals)?);

    let token_amount = balance.safe_mul(interest)?.safe_div(precision_decrease)?;

    Ok(token_amount)
}


pub fn get_strict_token_value(
    token_amount: i128,
    spot_decimals: u32,
    strict_price: &StrictOraclePrice
) -> NormalResult<i128> {
    if token_amount == 0 {
        return Ok(0);
    }

    let precision_decrease = (10_i128).pow(spot_decimals);

    let price = if token_amount > 0 { strict_price.min() } else { strict_price.max() };

    let token_with_price = token_amount.safe_mul(price.cast()?)?;

    if token_with_price < 0 {
        token_with_price.safe_div_floor(precision_decrease)
    } else {
        token_with_price.safe_div(precision_decrease)
    }
}

pub fn get_token_value(
    token_amount: i128,
    spot_decimals: u32,
    oracle_price: i64
) -> NormalResult<i128> {
    if token_amount == 0 {
        return Ok(0);
    }

    let precision_decrease = (10_i128).pow(spot_decimals);
    let token_with_oracle = token_amount.safe_mul(oracle_price.cast()?)?;

    if token_with_oracle < 0 {
        token_with_oracle.safe_div_floor(precision_decrease.abs())
    } else {
        token_with_oracle.safe_div(precision_decrease)
    }
}

File: ts-sdk/src/math/spotBalance.ts
import { MarketAccount, BalanceType, isVariant } from '../types';
import { BN } from '@coral-xyz/anchor';
import { ONE, TEN, ZERO } from '../constants/numericConstants';
import { OraclePriceData } from '../oracles/types';
import { divCeil } from './utils';
import { StrictOraclePrice } from '../oracles/strictOraclePrice';

/**
 * Calculates the balance of a given token amount including any accumulated interest. This
 * is the same as `SpotPosition.scaledBalance`.
 *
 * @param {BN} tokenAmount - the amount of tokens
 * @param {SpotMarketAccount} market - the spot market account
 * @param {BalanceType} balanceType - the balance type ('deposit' or 'borrow')
 * @return {BN} the calculated balance, scaled by `SPOT_MARKET_BALANCE_PRECISION`
 */
export function getBalance(
	tokenAmount: BN,
	market: MarketAccount,
	balanceType: BalanceType
): BN {
	const precisionIncrease = TEN.pow(new BN(19 - market.decimals));

	// const cumulativeInterest = isVariant(balanceType, 'deposit')
	// 	? market.cumulativeDepositInterest
	// 	: market.cumulativeBorrowInterest;

	let balance = tokenAmount.mul(precisionIncrease).div(new BN(1)); // cumulativeInterest

	if (!balance.eq(ZERO) && isVariant(balanceType, 'borrow')) {
		balance = balance.add(ONE);
	}

	return balance;
}

/**
 * Calculates the spot token amount including any accumulated interest.
 *
 * @param {BN} balanceAmount - The balance amount, typically from `SpotPosition.scaledBalance`
 * @param {SpotMarketAccount} market - The spot market account details
 * @param {BalanceType} balanceType - The balance type to be used for calculation
 * @returns {BN} The calculated token amount, scaled by `SpotMarketConfig.precision`
 */
export function getTokenAmount(
	balanceAmount: BN,
	market: MarketAccount,
	balanceType: BalanceType
): BN {
	const precisionDecrease = TEN.pow(new BN(19 - market.decimals));

	if (isVariant(balanceType, 'deposit')) {
		return balanceAmount
			.mul(new BN(1)) // market.cumulativeDepositInterest
			.div(precisionDecrease);
	} else {
		return divCeil(
			balanceAmount.mul(new BN(1)), // market.cumulativeBorrowInterest
			precisionDecrease
		);
	}
}

/**
 * Returns the signed (positive for deposit,negative for borrow) token amount based on the balance type.
 *
 * @param {BN} tokenAmount - The token amount to convert (from `getTokenAmount`)
 * @param {BalanceType} balanceType - The balance type to determine the sign of the token amount.
 * @returns {BN} - The signed token amount, scaled by `SpotMarketConfig.precision`
 */
export function getSignedTokenAmount(
	tokenAmount: BN,
	balanceType: BalanceType
): BN {
	if (isVariant(balanceType, 'deposit')) {
		return tokenAmount;
	} else {
		return tokenAmount.abs().neg();
	}
}

/**
 * Calculates the value of a given token amount using the worst of the provided oracle price and its TWAP.
 *
 * @param {BN} tokenAmount - The amount of tokens to calculate the value for (from `getTokenAmount`)
 * @param {number} spotDecimals - The number of decimals in the token.
 * @param {StrictOraclePrice} strictOraclePrice - Contains oracle price and 5min twap.
 * @return {BN} The calculated value of the given token amount, scaled by `PRICE_PRECISION`
 */
export function getStrictTokenValue(
	tokenAmount: BN,
	spotDecimals: number,
	strictOraclePrice: StrictOraclePrice
): BN {
	if (tokenAmount.eq(ZERO)) {
		return ZERO;
	}

	let price;
	if (tokenAmount.gte(ZERO)) {
		price = strictOraclePrice.min();
	} else {
		price = strictOraclePrice.max();
	}

	const precisionDecrease = TEN.pow(new BN(spotDecimals));

	return tokenAmount.mul(price).div(precisionDecrease);
}

/**
 * Calculates the value of a given token amount in relation to an oracle price data
 *
 * @param {BN} tokenAmount - The amount of tokens to calculate the value for (from `getTokenAmount`)
 * @param {number} spotDecimals - The number of decimal places of the token.
 * @param {OraclePriceData} oraclePriceData - The oracle price data (typically a token/USD oracle).
 * @return {BN} The value of the token based on the oracle, scaled by `PRICE_PRECISION`
 */
export function getTokenValue(
	tokenAmount: BN,
	spotDecimals: number,
	oraclePriceData: Pick<OraclePriceData, 'price'>
): BN {
	if (tokenAmount.eq(ZERO)) {
		return ZERO;
	}

	const precisionDecrease = TEN.pow(new BN(spotDecimals));

	return tokenAmount.mul(oraclePriceData.price).div(precisionDecrease);
}

// ...
